File: it.stefano.bls.domain/lib/src/main/java/led/concrete/LedDevice.java
package led.concrete;

import interf.ILedDevice;

public abstract class LedDevice implements ILedDevice {

	private boolean state = false;
	
	@Override
	public void on() {
		setState(true);
		doOn();
	}

	protected abstract void doOn();

	@Override
	public void off() {
		setState(false);
		doOff();
	}

	protected abstract void doOff();

	@Override
	public boolean isOn() {
		return state;
	}
	
	private void setState(boolean aState) {
		this.state = aState;
	}

}

File: it.stefano.bls.arduino/libraries/Led/LedDevice.cpp
#include "LedDevice.h"
#include <Arduino.h>

void LedDevice::on() {
  state = true;
  //Serial.println("LedDevice | on");
  doOn();
}
void LedDevice::off() {
  state = false;
  doOff();
}
bool LedDevice::isOn() {
  return state;
}